src/BalanceExpressionTrees.cpp

```diff
@@ -0,0 +1,522 @@
+#include "BalanceExpressionTrees.h"
+#include "IRMutator.h"
+#include "IREquality.h"
+#include "Scope.h"
+#include <algorithm>
+namespace Halide {
+namespace Internal {
+
+using std::vector;
+
+namespace {
+
+typedef std::pair<Expr, int> RootWeightPair;
+typedef std::map<Expr, int, IRDeepCompare> WeightedRoots;
+class ExprHeights : public IRVisitor {
+    using IRVisitor::visit;
+    std::map<Expr, int, IRDeepCompare> m;
+    Scope<int> var_heights;
+public:
+    // ExprHeights(Scope<int> var_heights) : var_heights(var_heights) {}
+    void clear() { m.clear(); }
+    void push(const std::string &name, int ht) {
+        var_heights.push(name, ht);
+    }
+    void pop(const std::string &name) {
+        var_heights.pop(name);
+    }
+    void push(Expr e) {
+        internal_assert(e.type().is_vector()) << "We are interested in the heights of only vector types\n";
```

Comment:
You are right. This is a vestige from the time that I had written this specifically for HVX and at that time I was trying to limit the scope of this.